src/endpoint.rs

```diff
@@ -297,4 +417,20 @@ impl Endpoint {
     pub(crate) fn bootstrap_nodes(&self) -> &[SocketAddr] {
         &self.bootstrap_nodes
     }
+
+    /// Returns the socket address of a peer that has connected to us
+    pub async fn next_incoming_connection(&mut self) -> Option<SocketAddr> {
+        self.connection_queue.lock().await.pop_front()
+    }
+
+    /// Returns a message received from another peer along with the source address
+    pub async fn next_incoming_message(&mut self) -> Option<(SocketAddr, Bytes)> {
+        self.message_queue.lock().await.pop_front()
+    }
```

Comment:
First of all, you don't need `&mut self` because the message queue is behind a `Mutex`. But more serious issue is that this seems like it forces us to busy-loop for incoming messages, wasting CPU cycles. Wonder why do we need the message queue at all. Why not just `recv` on the message channel receiver instead? This same point applies to `next_incoming_connection` and `next_disconnected_peer` too.